### ISSUE TITLE
Question: concurrency

### ISSUE THREAD

**Issue by harukizaemon** [jcupitt/libvips#561]
I'm running a bunch of batch image processing jobs using VIPS - mostly producing thumbnails and zoomable images (was DZI, then Zoomify, now Google) - but also some basic clipping and conversion from TIFF/PNG/JPEG <-> RAW files.

In the main, I probably have about 10 cores at my disposal. I'm trying to decide if that means I should run jobs in series and allow VIPS to use as many cores as I have available; or run a bunch in parallel and restrict VIPS to some subset of available cores.

Any suggestions greatly appreciated.

**Comment by jcupitt** [comment]
It depends where your bottleneck is.

If you are only doing simple stuff, like clip, shrink, dzsave and you are using formats which have rather "heavy" decode/encode libraries (like jpg or png), then you are probably limited by encode/decode speed. These libraries are always single-threaded. vips can run encode and decode in parallel, but that's all. In this case, running extra jobs, up to your physical (not hyperthreaded) core count will speed things up.

If you doing light work and using a light image format (like tiff), you will be limited by the speed of your disk channel. Running extra jobs will probably reduce throughput.

If you are CPU-limited (for example, filtering with a large radius, or using a very long pipeline), extra jobs can reduce throughput.

If you are hitting physical memory limits, extra jobs will hammer performance as you wander off into swap death.

tldr: running multiple vipses at once will often increase throughput (though not always), but don't start more jobs than you have physical cores.

**Comment by harukizaemon** [comment]
Brilliant. Thank you.
